cpp/src/io/json/read_json.cu

```diff
@@ -159,48 +177,89 @@ bool should_load_whole_source(json_reader_options const& opts, size_t source_siz
  * @param sources Data sources to read from
  * @param reader_opts JSON reader options with range offset and range size
  * @param stream CUDA stream used for device memory operations and kernel launches
- * @return Byte range for parsing
+ * @returns Data source owning buffer enclosing the bytes read
  */
-auto get_record_range_raw_input(host_span<std::unique_ptr<datasource>> sources,
-                                json_reader_options const& reader_opts,
-                                rmm::cuda_stream_view stream)
+datasource::owning_buffer<rmm::device_uvector<char>> get_record_range_raw_input(
+  host_span<std::unique_ptr<datasource>> sources,
+  json_reader_options const& reader_opts,
+  rmm::cuda_stream_view stream)
 {
-  auto buffer = ingest_raw_input(sources,
-                                 reader_opts.get_compression(),
-                                 reader_opts.get_byte_range_offset(),
-                                 reader_opts.get_byte_range_size(),
-                                 stream);
-  if (should_load_whole_source(reader_opts, sources[0]->size())) return buffer;
-  auto first_delim_pos =
-    reader_opts.get_byte_range_offset() == 0 ? 0 : find_first_delimiter(buffer, '\n', stream);
+  CUDF_FUNC_RANGE();
+  auto geometric_mean = [](double a, double b) { return std::sqrt(a * b); };
+
+  size_t const total_source_size            = sources_size(sources, 0, 0);
+  auto constexpr num_delimiter_chars        = 1;
+  auto const num_extra_delimiters           = num_delimiter_chars * (sources.size() - 1);
+  compression_type const reader_compression = reader_opts.get_compression();
+  size_t chunk_offset                       = reader_opts.get_byte_range_offset();
+  size_t chunk_size                         = reader_opts.get_byte_range_size();
+
+  CUDF_EXPECTS(total_source_size ? chunk_offset < total_source_size : !chunk_offset,
+               "Invalid offsetting");
+  auto should_load_all_sources = !chunk_size || chunk_size >= total_source_size - chunk_offset;
+  chunk_size =
+    should_load_all_sources ? total_source_size - chunk_offset + num_extra_delimiters : chunk_size;
+
+  // Some magic numbers
+  constexpr int num_subchunks        = 10;  // per chunk_size
+  constexpr size_t min_subchunk_size = 10000;
+  const int num_subchunks_prealloced = should_load_all_sources ? 0 : 3;
+  constexpr int compression_ratio    = 4;
```

Comment:
nit: This variable name makes me think we are setting a compression ratio, not guessing at possible ratio. Maybe something like `assumed_compression_ratio` or `estimated_compression_ratio` to drive home the fact that we don't know, but we need to guess for buffer sizes.

Reply:
Addressed.